src/test/lines.ts

```diff
@@ -0,0 +1,104 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {strMult} from "./utils";
+
+// Use classes here instead of interfaces because we want runtime type data
+export class Line { }
+export class CodeLine extends Line { constructor(public contents: string) { super(); } }
+export class MessageSubstitutionLine extends Line { constructor(public key: string, public message: string) { super(); } }
+
+export class ErrorLine extends Line { constructor(public startCol: number) { super(); } }
+export class MultilineErrorLine extends ErrorLine { constructor(startCol: number) { super(startCol); } }
+export class EndErrorLine extends ErrorLine {
+    constructor(startCol: number, public endCol: number, public message: string) { super(startCol); }
+}
+
+// example matches (between the quotes): 
+// "    ~~~~~~~~"
+const multilineErrorRegex = /^\s*(~+|~nil)$/;
+// "    ~~~~~~~~~   [some error message]"
+const endErrorRegex = /^\s*(~+|~nil)\s*\[(.+)\]\s*$/;
+// "[shortcut]: full messages goes here!!  "
+const messageSubstitutionRegex = /^\[([\w\-\_]+?)]: \s*(.+?)\s*$/;
+
+export const ZERO_LENGTH_ERROR = "~nil";
+
+/**
+ * Maps a line of text from a .linttest file to an appropriate Line object
+ */
+export function textToLineObj(text: string): Line {
+    const multilineErrorMatch = text.match(multilineErrorRegex);
+    if (multilineErrorMatch != null) {
+        const startErrorCol = text.indexOf("~");
+        return new MultilineErrorLine(startErrorCol);
+    }
+
+    const endErrorMatch = text.match(endErrorRegex);
+    if (endErrorMatch != null) {
+        const [, squiggles, message] = endErrorMatch;
+        const startErrorCol = text.indexOf("~");
+        const zeroLengthError = (squiggles === ZERO_LENGTH_ERROR);
+        const endErrorCol = zeroLengthError ? startErrorCol : text.lastIndexOf("~") + 1;
+        return new EndErrorLine(startErrorCol, endErrorCol, message);
+    }
+
+    const messageSubstitutionMatch = text.match(messageSubstitutionRegex);
+    if (messageSubstitutionMatch != null) {
+        const [, key, message] = messageSubstitutionMatch;
+        return new MessageSubstitutionLine(key, message);
+    }
+
+    // line doesn't match any syntax for error markup, so it's a line of code to be linted
+    return new CodeLine(text);
+}
+
+/**
+ * Maps an ErrorLine object to a matching line of text that could be in a .linttest file.
+ * This is almost the inverse of classifyLine.
+ * If you ran `createErrorString(classifyLine(someText))`, the whitespace in the result may be different than in someText
+ */
+export function lineObjToText(line: Line, code: string) {
```

Comment:
- add return type `: string`
- add JSDoc comments for parameters. it took me a few seconds to realize that `code` was the preceding line of code being linted here